rcl_logging_interface/test/test_get_logging_directory.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <rcpputils/filesystem_helper.hpp>
+#include <rcpputils/get_env.hpp>
+#include <rcutils/allocator.h>
+#include <rcutils/env.h>
+
+#include <string>
+
+#include "gtest/gtest.h"
+#include "rcl_logging_interface/rcl_logging_interface.h"
+
+// This is a helper class that resets an environment
+// variable when leaving scope
+class RestoreEnvVar
+{
+public:
+  explicit RestoreEnvVar(const std::string & name)
+  : name_(name),
+    value_(rcpputils::get_env_var(name.c_str()))
+  {
+  }
+
+  ~RestoreEnvVar()
+  {
+    if (!rcutils_set_env(name_.c_str(), value_.c_str())) {
+      std::cerr << "Failed to restore value of environment variable: " << name_ << std::endl;
+    }
+  }
+
+private:
+  const std::string name_;
+  const std::string value_;
+};
+
+TEST(test_logging_directory, directory)
+{
+  RestoreEnvVar home_var("HOME");
+  RestoreEnvVar userprofile_var("USERPROFILE");
+  ASSERT_EQ(true, rcutils_set_env("HOME", nullptr));
+  ASSERT_EQ(true, rcutils_set_env("USERPROFILE", nullptr));
+  ASSERT_EQ(true, rcutils_set_env("ROS2_LOG_DIR", nullptr));
+  ASSERT_EQ(true, rcutils_set_env("ROS2_HOME", nullptr));
+
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+
+  // Fails without any relevant env vars at all (HOME included)
+  const char * directory;
+  EXPECT_EQ(RCL_LOGGING_RET_ERROR, rcl_logging_get_logging_directory(&directory, allocator));
+
+  // Default case without ROS2_LOG_DIR or ROS2_HOME being set (but with HOME)
+  rcpputils::fs::path fake_home("/fake_home_dir");
+  ASSERT_EQ(true, rcutils_set_env("HOME", fake_home.string().c_str()));
+  rcpputils::fs::path default_dir = fake_home / ".ros" / "log";
+  EXPECT_EQ(RCL_LOGGING_RET_OK, rcl_logging_get_logging_directory(&directory, allocator));
+  EXPECT_STREQ(directory, default_dir.string().c_str());
```

Comment:
directory is being leaked (here and later)

Reply:
fixed in b50d081c9d24378b47b447969fadee97da2997ba